<a id = 'top'></a>
### HAND WRITTEN DIGITS RECOGNITION PROJECT
  - [PROJECT AIM AND DATA](#aim)
  - [USING KNN CLASSIFIER ](#knn)
  - [USING DECISION TREE CLASSIFIER](#dtree)
  - [USING RANDOM FOREST CLASSIFIER](#rforest)
  - [CONCLUSIONS](#conclusion)
---------------------------------------------------------------------[BACK TO TOP](#top)

In [1]:
# importing libraries
import numpy as np
import matplotlib.pyplot as plt
import tensorflow as tf
import sklearn
import time
import math      # not needed just in case
import random    # not needed just in case

<a id = 'aim'></a>
### PROJECT AIM AND DATA
#### AIM:- 
- developing a model that could detect a handwritten digits from 0 to 9
- and we will construct a supervised classification model using
   1. knn classifier
   2. decision tree classifier
   3. random forest classifier
   and address our problem
#### DATA
- we will be loading the digits dataset that 
- that stores 1797  9x9 pixels (flattened) grayscale images of handwritten digits from 0 to 9
- the size of dataset is 1797 x 81 
- which will be split into training and test datasets
- we will construct a model that can detect (classify any 9x9 flattend greyscale image of hand
  written digit ) into one of 0 to 9 or 1 to 10 numbers(classes)
---------------------------------------------------------------------[BACK TO TOP](#top)

In [2]:
# loading data for classification analysis 
# loading digits data set 
# and splitting it into test and train sets
from sklearn.datasets import load_digits
data = load_digits()
xtrain = data['data'][:1500,:]
ytrain = data['target'][:1500]
xtest = data['data'][1500:,:]
ytest = data['target'][1500:]
print(xtrain.shape,ytrain.shape)
print(xtest.shape,ytest.shape)

(1500, 64) (1500,)
(297, 64) (297,)


<a id = ''></a>
### KNN CLASSIFIER
---------------------------------------------------------------------[BACK TO TOP](#top)

In [3]:
from sklearn.neighbors import KNeighborsClassifier
cobject = KNeighborsClassifier(n_neighbors=5)
cobject.fit(xtrain,ytrain)
pred = cobject.predict(xtest)
from sklearn.metrics import classification_report,confusion_matrix
print(confusion_matrix(ytest,pred))
print(classification_report(ytest,pred))
# tune neighbors until you get highest test accuracy

[[26  0  0  0  1  0  0  0  0  0]
 [ 0 31  0  0  0  0  0  0  0  0]
 [ 0  0 27  0  0  0  0  0  0  0]
 [ 0  0  1 26  0  1  0  1  1  0]
 [ 0  0  0  0 31  0  0  1  1  0]
 [ 0  0  0  0  0 30  0  0  0  0]
 [ 0  0  0  0  0  0 30  0  0  0]
 [ 0  0  0  0  0  0  0 30  0  0]
 [ 0  2  0  0  0  0  0  1 25  0]
 [ 0  0  0  1  0  2  0  0  0 28]]
              precision    recall  f1-score   support

           0       1.00      0.96      0.98        27
           1       0.94      1.00      0.97        31
           2       0.96      1.00      0.98        27
           3       0.96      0.87      0.91        30
           4       0.97      0.94      0.95        33
           5       0.91      1.00      0.95        30
           6       1.00      1.00      1.00        30
           7       0.91      1.00      0.95        30
           8       0.93      0.89      0.91        28
           9       1.00      0.90      0.95        31

    accuracy                           0.96       297
   macro avg       

<a id = ''></a>
### DECISON TREE CLASSIFIER
---------------------------------------------------------------------[BACK TO TOP](#top)

In [4]:
from sklearn.tree import DecisionTreeClassifier
robject = DecisionTreeClassifier()
robject.fit(xtrain,ytrain)
pred = robject.predict(xtest)
from sklearn.metrics import classification_report,confusion_matrix
print(classification_report(ytest,pred))
print(confusion_matrix(ytest,pred))

              precision    recall  f1-score   support

           0       0.93      0.96      0.95        27
           1       0.72      0.58      0.64        31
           2       0.77      0.74      0.75        27
           3       0.64      0.53      0.58        30
           4       0.74      0.88      0.81        33
           5       0.86      0.83      0.85        30
           6       0.93      0.83      0.88        30
           7       0.62      0.80      0.70        30
           8       0.63      0.61      0.62        28
           9       0.78      0.81      0.79        31

    accuracy                           0.76       297
   macro avg       0.76      0.76      0.76       297
weighted avg       0.76      0.76      0.76       297

[[26  0  0  0  0  0  0  1  0  0]
 [ 0 18  3  1  3  0  0  0  5  1]
 [ 1  0 20  1  0  0  1  3  1  0]
 [ 0  1  2 16  1  3  0  5  2  0]
 [ 0  0  0  0 29  0  1  2  0  1]
 [ 0  0  0  1  3 25  0  0  0  1]
 [ 0  1  0  0  2  0 25  0  2  0]
 [ 0  0  0

<a id = ''></a>
### RANDOM FOREST CLASSIFIER
---------------------------------------------------------------------[BACK TO TOP](#top)

In [5]:
from sklearn.ensemble import RandomForestClassifier
robject = RandomForestClassifier(n_estimators = 10)
robject.fit(xtrain,ytrain)
pred = robject.predict(xtest)
from sklearn.metrics import classification_report,confusion_matrix
print(classification_report(ytest,pred))
print(confusion_matrix(ytest,pred))

              precision    recall  f1-score   support

           0       0.96      0.96      0.96        27
           1       0.78      0.94      0.85        31
           2       0.93      0.96      0.95        27
           3       0.95      0.63      0.76        30
           4       0.91      0.91      0.91        33
           5       0.87      0.90      0.89        30
           6       0.97      0.93      0.95        30
           7       0.88      0.97      0.92        30
           8       0.69      0.71      0.70        28
           9       0.87      0.84      0.85        31

    accuracy                           0.88       297
   macro avg       0.88      0.88      0.87       297
weighted avg       0.88      0.88      0.87       297

[[26  0  0  0  1  0  0  0  0  0]
 [ 1 29  0  0  0  0  0  0  0  1]
 [ 0  0 26  0  0  0  1  0  0  0]
 [ 0  3  1 19  0  1  0  0  6  0]
 [ 0  0  0  0 30  1  0  0  1  1]
 [ 0  1  0  1  0 27  0  0  0  1]
 [ 0  0  1  0  1  0 28  0  0  0]
 [ 0  0  0

<a id = ''></a>
### CONCLUSIONS
- even though these models have not yet been tuned to given their peak performance
  let us just compare their performance
                                    accuracy
  1. knn classifier            :  96%
  2. decision tree classifier  :  75% 
  3. random forest classifier  :  88%
- knn classifier outperformed the rest of the models
- these are test accuracies and not training 
---------------------------------------------------------------------[BACK TO TOP](#top)